### PR TITLE
mgr/dashboard: Explicitly set/change the device class of an OSD

### PR DESCRIPTION
**Pull request by ricardoasmarques**
**Screenshots:**
Screenshot from 2019-12-10 15-27-27

Screenshot from 2019-12-10 15-27-42

**Equivalent to:**
```
$ ceph osd crush get-device-class osd.2
$ ceph osd crush rm-device-class osd.2 
$ ceph osd crush set-device-class ssd osd.2
```

Fixes: https://tracker.ceph.com/issues/43197

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by ricardoasmarques** [comment]
I'm not sure if the menu entry (and modal title) should be '**Edit**' or '**CRUSH**'. 

@bk201 @LenzGr @votdev others, any thoughts?

**Comment by LenzGr** [comment]
Once question though - it does not seem to be possible to change the device class on multiple OSDs at once (the "CRUSH" menu item is disabled): is this intentional?

**Comment by ricardoasmarques** [comment]
> Once question though - it does not seem to be possible to change the device class on multiple OSDs at once (the "CRUSH" menu item is disabled): is this intentional?

It's intentional, but if we find it's useful we can add the support for that (maybe something for a separate PR).

**Comment by votdev** [comment]
Have you considered to use [FormModalComponent](https://github.com/ceph/ceph/blob/master/src/pybind/mgr/dashboard/frontend/src/app/shared/components/form-modal/form-modal.component.ts#L22) as a replacement for your `OsdCrushModalComponent` dialog? You can find an example [here](https://github.com/ceph/ceph/blob/master/src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-directories/cephfs-directories.component.ts#L280).

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts [hunk 1]
```diff
@@ -302,6 +331,17 @@ export class OsdListComponent implements OnInit {
     });
   }
 
+  crushAction() {
+    const selectedOsd = this.osds.filter((o) => o.id === this.selection.first().id).pop();
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts]: Alternatively use lodash.
```suggestion
    const selectedOsd = _.filter(this.osds, ['id', this.selection.first().id]).pop();
```

**Comment by ricardoasmarques** [comment]
> Have you considered to use [FormModalComponent](https://github.com/ceph/ceph/blob/master/src/pybind/mgr/dashboard/frontend/src/app/shared/components/form-modal/form-modal.component.ts#L22) i've created some time ago as a replacement for your `OsdCrushModalComponent` dialog? This will save you many lines of code. You can find an example [here](https://github.com/ceph/ceph/blob/master/src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-directories/cephfs-directories.component.ts#L280). Right at the moment the `FormModalComponent` only supports input fields, but can be enhanced easily (Stephan is doing that for example in one of his PR).

I'm now using `FormModalComponent`, thanaks.

**Comment by ricardoasmarques** [comment]
> Have you considered to use [FormModalComponent](https://github.com/ceph/ceph/blob/master/src/pybind/mgr/dashboard/frontend/src/app/shared/components/form-modal/form-modal.component.ts#L22) i've created some time ago as a replacement for your `OsdCrushModalComponent` dialog? This will save you many lines of code. You can find an example [here](https://github.com/ceph/ceph/blob/master/src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-directories/cephfs-directories.component.ts#L280). Right at the moment the `FormModalComponent` only supports input fields, but can be enhanced easily (Stephan is doing that for example in one of his PR).

I'm now using `FormModalComponent`, thanks.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts (hunk 1 above)
Comment (ricardoasmarques) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts]: Done

Comment (ricardoasmarques) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts]: Done

**Comment by ricardoasmarques** [comment]
> I'd suggest, at least, "Device Class" instead, but probably "Edit" would be better as it would serve also as a placeholder for any further OSD action. IMHO, that menu has a lot of sub-entries (and a bit unbalanced): Create, Scrub, Deep Scrub, Reweigh, Mark *, Purge, Destroy. Compare that to the regular Create, Edit, Delete that have most other components (Pools, RBD, RGW).

I've renamed it to "Edit". Thanks for the feedback.

**Comment by callithea** [comment]
- [pkg build](https://shaman.ceph.com/builds/ceph/wip-laura-testing-32150/) [WIP, had to re-trigger it because some builds were missing for the previous one]
- QA run [ ]

**Comment by callithea** [comment]
jenkins test dashboard backend

**Comment by callithea** [comment]
jenkins test dashboard backend

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-12-21_03:01:01-rados-wip-kefu-testing-2019-12-21-0133-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-12-23_16:20:31-rados-wip-kefu-testing-2019-12-21-0133-distro-basic-smithi/

failures tracked by

- https://tracker.ceph.com/issues/43396
- https://tracker.ceph.com/issues/43415
- https://tracker.ceph.com/issues/43414
- https://tracker.ceph.com/issues/43416
- https://tracker.ceph.com/issues/43394

**Comment by tchaikov** [comment]
jenkins test dashboard backend

**Pull request merged**